### DISCUSSION TITLE
[Compilers] How do I hash bytecode for create2 address in solidity

### DISCUSSION THREAD

**Discussion by ghost** [zkSync-Community-Hub/zksync-developers#290] [Compilers] [answered]
### Which compiler are you using?

- [X] zksolc
- [ ] zkvyper

### Which version of Solidity or Vyper?

latest

### zksolc/zkvyper version?

latest

### Explain the issue

Would like to know how to retrieve hashBytecode value in solidity.

There is a documentation for js https://docs.zksync.io/build/sdks/js/zksync-ethers/utils.html#hashbytecode however there is no documentation for solidity.

**Comment by dutterbutter** [comment, marked as answer]
Hey 👋 Please review the create2 documentation which can be found here:
- https://docs.zksync.io/build/developer-reference/differences-with-ethereum.html#create-create2

Example,
```
bytes memory bytecode = type(MyContract).creationCode;
assembly {
    addr := create2(0, add(bytecode, 32), mload(bytecode), salt)
}

```

**Comment by lysumeduso** [comment]
In Solidity, you can compute the bytecode hash using the built-in keccak256 function.
// Simply pass the bytecode as an argument to keccak256 to get the hash.
 Example: bytes32 hashBytecode = keccak256(type(ContractName).creationCode);
 Replace "ContractName" with the name of your contract.
 This will give you the hash of the bytecode.

**Comment by fitzgibbonsdustin** [comment]
In Solidity, you can calculate the hash of a contract's bytecode using the keccak256 function. Here's a simple example:
 SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

contract BytecodeHash 
     Function to calculate the hash of the bytecode
    function calculateBytecodeHash() public pure returns (bytes32) 
         Get the bytecode of the contract
        bytes memory bytecode = type(BytecodeHash).runtimeCode;
         Calculate the hash using keccak256
        bytes32 hash = keccak256(bytecode);
        return hash;

This contract provides a calculateBytecodeHash function that returns the hash of its own bytecode. You can call this function to retrieve the hash value.

**Comment by phicurby** [comment]
To retrieve the hashBytecode value in Solidity, you can use the hashBytecode function provided by the zksync-ethers library. This function calculates the hash of the provided bytecode. Make sure to include the zksync-ethers library in your Solidity project and call the hashBytecode function with the bytecode of the contract you want to hash.

**Comment by jwitte3** [comment]
To retrieve the hashBytecode value in Solidity, you can use the zkSyncUtils library. Here's a simple example:

import { zkSyncUtils } from "zksync";

contract MyContract {
    function getHashBytecode() public pure returns (bytes32) {
        return zkSyncUtils.hashBytecode(bytecode);
    }
}
Make sure to replace bytecode with the bytecode of the contract you want to hash.

**Comment by davehutchinsv** [comment]
The user is seeking guidance on how to retrieve the hashBytecode value in Solidity for zkSync. While there is documentation for retrieving the hashBytecode value in JavaScript, there is no documentation currently available for Solidity. Further investigation and research may be needed to determine a solution to this issue.
